compiler/rustc_trait_selection/src/solve/inspect/build.rs

```diff
@@ -216,6 +216,34 @@ impl<'tcx> WipGoalEvaluationStep<'tcx> {
         }
         inspect::GoalEvaluationStep { instantiated_goal: self.instantiated_goal, evaluation }
     }
+
+    // Returns all added goals from this scope and all containing scopes.
+    fn all_added_goals(
+        &self,
+    ) -> impl Iterator<Item = (GoalSource, Goal<'tcx, ty::Predicate<'tcx>>)> + '_ {
+        std::iter::from_coroutine(
```

Comment:
Chose to use a generator here but I could very well write this using a flat map or something, if you'd prefer. Or just return a vec.